### PR TITLE
[walreplication] Mask sequences and re-add sequence checks to gp_replica_check test

### PR DESCRIPTION
**Pull request by danielgustafsson**
Fix masking of `last_value` and `log_cnt` in the sequences to account for the pre-logging, and re-add sequences to `gp_replica_check` tests.

This also includes a few small patches to `gp_replica_check` that I found helpful, see individual commit messages for info.

@hlinnaka

### CODE REVIEW

File: src/backend/commands/sequence.c [hunk 1]
```diff
@@ -1731,13 +1732,51 @@ cdb_sequence_nextval_server(Oid    tablespaceid,
...
+		MemSet(&seqtup->log_cnt, 0, sizeof(int64));
+	}
+}
+
 /*
  * Mask a Sequence page before performing consistency checks on it.
  */
 void
-seq_mask(char *page, BlockNumber blkno)
+seq_mask(char *pagedata, BlockNumber blkno)
```
Comment (ashwinstar) [inline src/backend/commands/sequence.c]: This would make a diff from upstream do we really wish to do it ?

File: src/backend/commands/sequence.c [hunk 2]
```diff
@@ -1731,13 +1732,51 @@ cdb_sequence_nextval_server(Oid    tablespaceid,
     cdb_sequence_relation_term(seqrel);
 }                               /* cdb_sequence_server_nextval */
 
+/*
+ * Mask last_value and log_cnt for consistency checking
+ *
+ * To avoid logging every fetch from a sequence, SEQ_LOG_VALS are pre-logged
```
Comment (ashwinstar) [inline src/backend/commands/sequence.c]: Curious to know, why upstream doesn't need this masking. Has Sequence behavior different in later versions of Postgres or this is missing in upstream as well and needs to be submitted ?

**Comment by danielgustafsson** [comment]
> On 05 Jan 2018, at 13:24, Ashwin Agrawal <notifications@github.com> wrote:

> Curious to know, why upstream doesn't need this masking. Has Sequence behavior different in later versions of Postgres or this is missing in upstream as well and needs to be submitted ?

I’m not really sure why actually.

**Comment by ashwinstar** [comment]
On Fri, Jan 5, 2018 at 4:48 AM, Daniel Gustafsson <notifications@github.com>
wrote:

> > On 05 Jan 2018, at 13:24, Ashwin Agrawal <notifications@github.com>
> wrote:
>
> > Curious to know, why upstream doesn't need this masking. Has Sequence
> behavior different in later versions of Postgres or this is missing in
> upstream as well and needs to be submitted ?
>
> I’m not really sure why actually.


Upstream seems to still WAL the same way, so should be able to repro the
problem with Sequence and wal_consistency_check enabled.

**Comment by danielgustafsson** [comment]
> On 05 Jan 2018, at 13:59, Ashwin Agrawal <notifications@github.com> wrote:
> 
> On Fri, Jan 5, 2018 at 4:48 AM, Daniel Gustafsson <notifications@github.com>
> wrote:
> 
> > > On 05 Jan 2018, at 13:24, Ashwin Agrawal <notifications@github.com>
> > wrote:
> >
> > > Curious to know, why upstream doesn't need this masking. Has Sequence
> > behavior different in later versions of Postgres or this is missing in
> > upstream as well and needs to be submitted ?
> >
> > I’m not really sure why actually.
> 
> Upstream seems to still WAL the same way, so should be able to repro the
> problem with Sequence and wal_consistency_check enabled.

I haven’t had the time to test this, but I have a theory. wal_consistency_checking in upstream compares the replayed buffer with the stored image, while gp_replica_check compares the QD buffer page with the standby buffer page (or segment/mirror etc). This means that the replayed last_value and log_cnt will match what was stored in the WAL image. I think that can explain why this masking isn’t needed in upstream. Either that or I’m totally off.

**Comment by danielgustafsson** [comment]
> On 05 Jan 2018, at 13:24, Ashwin Agrawal <notifications@github.com> wrote:
> 
> @ashwinstar commented on this pull request.
> 
> In src/backend/commands/sequence.c:
> 
> >  /*
>   * Mask a Sequence page before performing consistency checks on it.
>   */
>  void
> -seq_mask(char *page, BlockNumber blkno)
> +seq_mask(char *pagedata, BlockNumber blkno)
> 
> This would make a diff from upstream do we really wish to do it ?

Good point, even though I prefer this version (which is also what most XXX_mask() functions in upstream looks like), avoiding a diff is a good thing. Will revert this back to upstream coding before merging.

**Comment by ashwinstar** [comment]
On Fri, Jan 5, 2018 at 6:27 AM, Daniel Gustafsson <notifications@github.com>
wrote:

> > On 05 Jan 2018, at 13:59, Ashwin Agrawal <notifications@github.com>
> wrote:
> >
> > On Fri, Jan 5, 2018 at 4:48 AM, Daniel Gustafsson <
> notifications@github.com>
> > wrote:
> >
> > > > On 05 Jan 2018, at 13:24, Ashwin Agrawal <notifications@github.com>
> > > wrote:
> > >
> > > > Curious to know, why upstream doesn't need this masking. Has Sequence
> > > behavior different in later versions of Postgres or this is missing in
> > > upstream as well and needs to be submitted ?
> > >
> > > I’m not really sure why actually.
> >
> > Upstream seems to still WAL the same way, so should be able to repro the
> > problem with Sequence and wal_consistency_check enabled.
>
> I haven’t had the time to test this, but I have a theory.
> wal_consistency_checking in upstream compares the replayed buffer with the
> stored image, while gp_replica_check compares the QD buffer page with the
> standby buffer page (or segment/mirror etc). This means that the replayed
> last_value and log_cnt will match what was stored in the WAL image. I think
> that can explain why this masking isn’t needed in upstream. Either that or
> I’m totally off.


It says when WAL is generated, tuple (means current page) doesn't match the
effect replay would have. So for wal_consistency_check when it captures the
page image while writing wal has to differ from what gets relayed and then
compared with.

We should get this in, then can later see if really needed for upstream or
not.

**Comment by danielgustafsson** [comment]
Pushed to the walreplication branch to not block that work, figuring out if upstream needs this can be a parallell activity.

**Pull request closed without merging**